From 7e7ba604125d17679441ae4cad76832b39d53c80 Mon Sep 17 00:00:00 2001
From: Natalia Kovalenko <natasha.k@variscite.com>
Date: Fri, 5 Jul 2024 16:06:43 +0200
Subject: [PATCH] mxm_wifiex: Add he_gi to nl80211_he_gi conversion

To Fix WARN_ON_ONCE() in the cfg80211_calculate_bitrate_he():

'WARNING: CPU: 0 PID: 565 at net/wireless/util.c:1384 cfg80211_calculate_bitrate_he+0x128/0x148 [cfg80211]'

add missing conversion to enum nl80211_he_gi when assigning to
he_gi.

derived from commit 26246bf60afa613272156fa268e4ff99f5d810ae

Signed-off-by: Natalia Kovalenko <natasha.k@variscite.com>
---
 .../wlan_src/mlinux/moal_sta_cfg80211.c       | 37 ++++++++++++++++++-
 1 file changed, 35 insertions(+), 2 deletions(-)

diff --git a/mxm_wifiex/wlan_src/mlinux/moal_sta_cfg80211.c b/mxm_wifiex/wlan_src/mlinux/moal_sta_cfg80211.c
index f13f641..0f7cda3 100644
--- a/mxm_wifiex/wlan_src/mlinux/moal_sta_cfg80211.c
+++ b/mxm_wifiex/wlan_src/mlinux/moal_sta_cfg80211.c
@@ -3304,6 +3304,39 @@ done:
 	return ret;
 }
 
+#if CFG80211_VERSION_CODE >= KERNEL_VERSION(5, 1, 18)
+/**
+ * @brief This function convert he_gi to nl80211_he_gi
+ *
+ * @param he_gi  0/1/2/3
+ *
+ *
+ * @return  0: NL80211_RATE_INFO_HE_GI_0_8
+ *          1: NL80211_RATE_INFO_HE_GI_1_6
+ *          2: NL80211_RATE_INFO_HE_GI_3_2
+ */
+static enum nl80211_he_gi woal_he_gi_to_nl80211_he_gi(u8 he_gi)
+{
+	enum nl80211_he_gi cfg_he_gi;
+
+	switch (he_gi) {
+	case 3:
+		cfg_he_gi = NL80211_RATE_INFO_HE_GI_3_2;
+		break;
+	case 2:
+		cfg_he_gi = NL80211_RATE_INFO_HE_GI_1_6;
+		break;
+	case 0:
+	case 1:
+	default:
+		cfg_he_gi = NL80211_RATE_INFO_HE_GI_0_8;
+		break;
+	}
+
+	return cfg_he_gi;
+}
+#endif
+
 /**
  * @brief Request the driver to fill the tx/rx rate info
  *
@@ -3380,7 +3413,7 @@ static void woal_cfg80211_fill_rate_info(moal_private *priv,
 			sinfo->txrate.flags = RATE_INFO_FLAGS_HE_MCS;
 			sinfo->txrate.nss = rate->param.data_rate.tx_nss + 1;
 			sinfo->txrate.mcs = rate->param.data_rate.tx_mcs_index;
-			sinfo->txrate.he_gi = rate->param.data_rate.tx_ht_gi;
+			sinfo->txrate.he_gi = woal_he_gi_to_nl80211_he_gi(rate->param.data_rate.tx_ht_gi);
 			if (rate->param.data_rate.tx_ht_bw == MLAN_VHT_BW80)
 				sinfo->txrate.bw = RATE_INFO_BW_80;
 			else if (rate->param.data_rate.tx_ht_bw == MLAN_HT_BW40)
@@ -3427,7 +3460,7 @@ static void woal_cfg80211_fill_rate_info(moal_private *priv,
 			sinfo->rxrate.flags = RATE_INFO_FLAGS_HE_MCS;
 			sinfo->rxrate.nss = rate->param.data_rate.rx_nss + 1;
 			sinfo->rxrate.mcs = rate->param.data_rate.rx_mcs_index;
-			sinfo->rxrate.he_gi = rate->param.data_rate.rx_ht_gi;
+			sinfo->rxrate.he_gi = woal_he_gi_to_nl80211_he_gi(rate->param.data_rate.rx_ht_gi);
 			if (rate->param.data_rate.rx_ht_bw == MLAN_VHT_BW80)
 				sinfo->rxrate.bw = RATE_INFO_BW_80;
 			else if (rate->param.data_rate.rx_ht_bw == MLAN_HT_BW40)
-- 
2.34.1
